File: c++/MaxFlow.cpp
#include <algorithm>
#include <limits>
#include <vector>

using namespace std;

template<typename C, typename E> class PushRelabelMaxFlowGraph {
  struct DListEntry {
    int prev, next;
  };
  struct Edge {
    C f;
    const C c;
    const int v, e;
    Edge (C _c, int _v, int _e) :
      f (0),
      c (_c),
      v (_v),
      e (_e)
    {
    }
  };
  const int n;
  int maxh;
  vector<vector<Edge> > edges;
  vector<int> current, h, nl, gc;
  vector<E> e;
  vector<DListEntry> dl;
  void addLink (int u, int v) {
    dl[u].next = v;
    dl[v].prev = u;
  }
  void insert (int ht, int v) {
    addLink (dl[n + ht].prev, v);
    addLink (v, n + ht);
  }
  void remove (int i) {
    addLink (dl[i].prev, dl[i].next);
  }
  void initPreflow () {
    for (int i = n; i < (int) dl.size (); ++i) {
      dl[i].prev = dl[i].next = i;
    }
    h[0] = n;
    for (auto& p : edges[0]) {
      const auto i = p.v;
      if (!e[i] && i && i < n - 1) {
        insert (0, i);
      }
      e[i] += p.f = p.c;
      edges[p.v][p.e].f = -p.c;
    }
    gc[0] = n - 1;
  }
  void push (int i, Edge &edge) {
    const auto j = edge.v;
    E d = edge.c - edge.f;
    if (d > e[i]) d = e[i];
    edge.f += d;
    edges[j][edge.e].f = -edge.f;
    e[i] -= d;
    e[j] += d;
  }
  void lift (int i) {
    int m = numeric_limits<int>::max();
    for (const auto& p : edges[i]) {
      if (p.f < p.c && m > h[p.v]) {
        m = h[p.v];
      }
    }
    h[i] = m + 1;
  }
  void discharge (int i) {
    while (e[i] > 0) {
      if (current[i] == (int) edges[i].size ()) {
        current[i] = 0;
        if (!(--gc[maxh]) && maxh && maxh < n) {
          remove (i);
          h[i] = n + 1;
          for (int k = 1; k < n; ++k) {
            if (h[k] > maxh && h[k] <= n) {
              --gc[h[k]];
              h[k] = n + 1;
            }
          }
          maxh = nl[maxh];
          break;
        } else {
          lift (i);
          remove (i);
          if (h[i] < n) {
            ++gc[h[i]];
            insert (h[i], i);
            nl[h[i]] = (dl[maxh + n].next >= n) ? nl[maxh] : maxh;
            maxh = h[i];
          } else {
            if (dl[maxh + n].next >= n) {
              maxh = nl[maxh];
            }
            break;
          }
        }
      } else {
        auto &p = edges[i][current[i]];
        const auto j = p.v;
        if (h[i] == h[j] + 1 && p.f < p.c) {
          const bool aj = e[j] > 0;
          push (i, p);
          if (!aj && j && j < n - 1 && e[j] > 0) {
            insert (maxh - 1, j);
            if (nl[maxh] != maxh - 1) {
              nl[maxh-1] = nl[maxh];
              nl[maxh] = maxh - 1;
            }
          }
          if (e[i] <= 0) {
            remove (i);
            if (dl[maxh + n].next >= n) {
              maxh = nl[maxh];
            }
          }
        } else {
          ++current[i];
        }
      }
    }
  }
  public:
  void addEdge (int i, int j, C w1, C w2) {
    const auto ei = (int) edges[i].size (), ej = (int) edges[j].size ();
    edges[i].emplace_back (w1, j, ej);
    edges[j].emplace_back (w2, i, ei);
  }
  //0 - sink, (n - 1) - target
  PushRelabelMaxFlowGraph (int _n) :
    n (_n),
    edges (n),
    current (n),
    h (n), nl (n, -1), gc (n, 0),
    e (n),
    dl (2 * n)
  {
  }
  E maxFlow () {
    initPreflow ();
    for (int i = 1; i < n - 1; ++i) {
      current[i] = 0;
    }
    maxh = 0;
    while (maxh >= 0) {
      int v = dl[maxh + n].next;
      if (v >= n) {
        break;
      }
      discharge (v);
    }
    return e[n-1];
  }
};

File: rust/src/maxflow.rs
type C = i32;
type E = i32;

#[derive(Clone, Debug)]
struct DListEntry {
  prev: usize,
  next: usize,
}

#[derive(Clone, Debug)]
struct Edge {
  v: usize,
  e: usize,
  f: C,
  c: C,
}

impl Edge {
  fn new(v: usize, e: usize, c: C) -> Self {
    Edge { v, e, f: 0, c }
  }
  fn zero() -> Self {
    Self::new(0, 0, 0)
  }
}

pub struct PushRelabelMaxFlowGraph {
  edges: Vec<Vec<Edge>>,
  current: Vec<usize>,
  h: Vec<i32>,
  nl: Vec<i32>,
  e: Vec<E>,
  dl: Vec<DListEntry>,
  gc: Vec<i32>,
  n: usize,
  maxh: i32,
}

impl PushRelabelMaxFlowGraph {
  pub fn new(n: usize) -> Self {
    PushRelabelMaxFlowGraph {
      edges: vec![vec![Edge::zero(); 0]; n],
      current: vec![std::usize::MAX; n],
      h: vec![0; n],
      nl: vec![-1; n],
      e: vec![0; n],
      dl: vec![DListEntry { prev: 0, next: 0 }; 2 * n],
      gc: vec![0; n],
      n,
      maxh: 0,
    }
  }
  fn add_link(&mut self, u: usize, v: usize) {
    self.dl[u].next = v;
    self.dl[v].prev = u;
  }
  fn insert(&mut self, ht: usize, v: usize) {
    self.add_link(self.dl[self.n + ht].prev, v);
    self.add_link(v, self.n + ht);
  }
  fn remove(&mut self, i: usize) {
    self.add_link(self.dl[i].prev, self.dl[i].next);
  }
  fn init_preflow(&mut self) {
    let n = self.n;
    for i in n..2 * n {
      self.dl[i].prev = i;
      self.dl[i].next = i;
    }
    self.h[0] = n as i32;
    for k in 0..self.edges[0].len() {
      let p = &mut self.edges[0][k];
      let i = p.v;
      let c = p.c;
      let e = p.e;
      p.f = c;
      if i > 0 && i < n - 1 && 0 == self.e[i] {
        self.insert(0, i);
      }
      self.e[i] += c;
      self.edges[i][e].f = -c;
    }
    self.gc[0] = (n - 1) as i32;
  }
  fn push(&mut self, i: usize, e: usize) {
    let p = &mut self.edges[i][e];
    let j = p.v;
    let d = std::cmp::min(self.e[i], p.c - p.f);
    let pe = p.e;
    p.f += d;
    self.edges[j][pe].f = -p.f;
    self.e[i] -= d;
    self.e[j] += d;
  }
  fn lift(&mut self, i: usize) {
    let mut m = i32::MAX;
    for p in &self.edges[i] {
      if p.f < p.c && m > self.h[p.v] {
        m = self.h[p.v];
      }
    }
    self.h[i] = m + 1;
  }
  fn discharge(&mut self, i: usize) {
    let n = self.n;
    while self.e[i] > 0 {
      if self.current[i] >= self.edges[i].len() {
        self.current[i] = 0;
        self.gc[self.maxh as usize] -= 1;
        if 0 == self.gc[self.maxh as usize] && self.maxh > 0 && self.maxh < (n as i32) {
          self.remove(i);
          self.h[i] = (n + 1) as i32;
          for k in 1..n {
            if self.h[k] > (self.maxh as i32) && self.h[k] <= (n as i32) {
              self.gc[self.h[k] as usize] -= 1;
              self.h[k] = (n + 1) as i32;
            }
          }
          self.maxh = self.nl[self.maxh as usize];
          break;
        } else {
          let maxh = self.maxh;
          self.lift(i);
          self.remove(i);
          if self.h[i] < (n as i32) {
            self.gc[self.h[i] as usize] += 1;
            self.insert(self.h[i] as usize, i);
            self.nl[self.h[i] as usize] = if self.dl[maxh as usize + n].next >= n {
              self.nl[maxh as usize]
            } else {
              maxh
            };
            self.maxh = self.h[i];
          } else {
            if self.dl[maxh as usize + n].next >= n {
              self.maxh = self.nl[maxh as usize];
            }
            break;
          }
        }
      } else {
        let e = self.current[i];
        let p = &self.edges[i][e];
        let j = p.v;
        let pf = p.f;
        let pc = p.c;
        if self.h[i] == self.h[j] + 1 && pf < pc {
          let aj = self.e[j] <= 0;
          self.push(i, e);
          if aj && j > 0 && j < n - 1 && self.e[j] > 0 {
            self.insert((self.maxh - 1) as usize, j);
            if self.nl[self.maxh as usize] != self.maxh - 1 {
              self.nl[(self.maxh - 1) as usize] = self.nl[self.maxh as usize];
              self.nl[self.maxh as usize] = self.maxh - 1;
            }
          }
          if self.e[i] <= 0 {
            self.remove(i);
            if self.dl[(self.maxh as usize) + n].next >= n {
              self.maxh = self.nl[self.maxh as usize];
            }
          }
        } else {
          self.current[i] += 1;
        }
      }
    }
  }
  pub fn add_edge(&mut self, i: usize, j: usize, w1: C, w2: C) {
    let ei = self.edges[i].len();
    let ej = self.edges[j].len();
    self.edges[i].push(Edge::new(j, ej, w1));
    self.edges[j].push(Edge::new(i, ei, w2));
  }
  pub fn max_flow(&mut self) -> E {
    let n = self.n;
    self.init_preflow();
    for i in 1..n - 1 {
      self.current[i] = 0;
    }
    self.maxh = 0;
    while self.maxh >= 0 {
      let v = self.dl[(self.maxh as usize) + n].next;
      if v >= n {
        break;
      }
      self.discharge(v);
    }
    self.e[n - 1]
  }
}
